runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/BoundedTrieData.java

```diff
@@ -0,0 +1,593 @@
+
+  /**
+   * Creates a {@link BoundedTrieData} instance.
+   *
+   * @param singleton The singleton path (optional).
+   * @param root The root node of the trie (optional).
+   * @param bound The maximum size of the trie.
+   * @throws IllegalArgumentException If both or neither of {@code singleton} and {@code root} are
+   *     specified or both are null, or if {@code bound} is less than 1.
+   */
+  public static BoundedTrieData create(
+      @Nullable List<String> singleton, @Nullable BoundedTrieNode root, int bound) {
+    Preconditions.checkArgument(
+        (singleton == null ^ root == null),
+        "Either and only one of singleton or root must be specified.");
+    Preconditions.checkArgument(bound >= 1, "Bound must be at least 1.");
+    return new AutoValue_BoundedTrieData(
+        Optional.ofNullable(singleton), Optional.ofNullable(root), bound);
+  }
+
+  /**
+   * Creates a {@link BoundedTrieData} instance from a {@link BoundedTrieNode} with the default
+   * bound.
+   *
+   * @param root The root node of the trie.
+   */
+  public static BoundedTrieData create(@Nonnull BoundedTrieNode root) {
+    return create(null, root, DEFAULT_BOUND);
+  }
+
+  /**
+   * Creates a {@link BoundedTrieData} instance from a singleton path with the default bound.
+   *
+   * @param singleton The singleton path.
+   */
+  public static BoundedTrieData create(@Nonnull List<String> singleton) {
+    return create(singleton, null, DEFAULT_BOUND);
+  }
+
+  /** Converts this {@link BoundedTrieData} to its proto {@link BoundedTrie}. */
+  public BoundedTrie toProto() {
+    BoundedTrie.Builder builder = BoundedTrie.newBuilder();
+    builder.setBound(bound());
+    singleton().ifPresent(builder::addAllSingleton);
+    root().ifPresent(r -> builder.setRoot(r.toProto()));
+    return builder.build();
+  }
+
+  /** Creates a {@link BoundedTrieData} instance from its proto {@link BoundedTrie}. */
+  public static BoundedTrieData fromProto(BoundedTrie proto) {
+    List<String> singleton = proto.getSingletonList();
+    BoundedTrieNode root = proto.hasRoot() ? BoundedTrieNode.fromProto(proto.getRoot()) : null;
+    return create(singleton, root, proto.getBound());
+  }
+
+  /** Returns this {@link BoundedTrieData} as a {@link BoundedTrieNode}. */
+  public BoundedTrieNode asTrie() {
+    return root()
+        .orElseGet(
+            () -> {
+              BoundedTrieNode newRoot = new BoundedTrieNode();
+              singleton().ifPresent(newRoot::add);
+              return newRoot;
+            });
+  }
+
+  /** Returns a new {@link BoundedTrieData} instance that is a deep copy of this instance. */
+  public BoundedTrieData getCumulative() {
+    return root().isPresent()
+        ? create(null, new BoundedTrieNode(root().get()), bound())
+        : create(singleton().get(), null, bound());
+  }
+
+  /** Extracts the data from this {@link BoundedTrieData} as a {@link BoundedTrieResult}. */
+  public BoundedTrieResult getBoundedTrieResult() {
```

Comment:
This kinda seem redundant, BoundedTrieResult just a wrapper on the result. I don't remember from StringSet changes if XYZResult classes are useful in e2e metric plumbing or not. When I do that and this seem like a wrapper only I will get rid of it so devs don't have to do .getBoundedTrieResult.getResult. Unless there is objection around differing from standard pattern